File: hash/go/imagehash/phash_simple.go
package imagehash

import (
	"fmt"
	"image"

	"github.com/wmetcalf/rosetta-image-hash/go/imagehash/internal/dct"
	"github.com/wmetcalf/rosetta-image-hash/go/imagehash/internal/imgrgb"
	"github.com/wmetcalf/rosetta-image-hash/go/imagehash/internal/lanczos"
)

// PHashSimple with default highfreqFactor=4.
func PHashSimple(img image.Image, hashSize int) (Hash, error) {
	return PHashSimpleWithFactor(img, hashSize, 4)
}

// PHashSimpleWithFactor computes phash_simple.
//
// Unlike PHash (2-D DCT, median threshold), phash_simple applies a 1-D
// DCT row-wise only and uses the mean as the threshold. It also takes a
// different block slice — columns 1 through N (skipping the DC column 0).
//
// This matches the Python imagehash 4.3.2 implementation of phash_simple:
//
//	dct = scipy.fftpack.dct(pixels)     # 1-D DCT along each row
//	dctlowfreq = dct[:hash_size, 1:hash_size+1]
//	avg = dctlowfreq.mean()
//	diff = dctlowfreq > avg
//
// scipy.fftpack.dct(x) for a 2-D array x applies along the last axis (rows).
func PHashSimpleWithFactor(img image.Image, hashSize, highfreqFactor int) (Hash, error) {
	if hashSize < 2 {
		return Hash{}, fmt.Errorf("hashSize must be >= 2, got %d", hashSize)
	}
	imgSize := hashSize * highfreqFactor

	rgb := imgrgb.ToRGB(img)
	gray := rgbToGray(rgb)
	resized := lanczos.Resize(gray, imgSize, imgSize)

	// Apply 1-D DCT row-wise.
	// Uses the same DCT1D that PHash uses (Makhoul FFT trick for power-of-2 sizes,
	// which preserves exact zeros and matches scipy exactly).
	dctRows := make([][]float64, imgSize)
	for y := 0; y < imgSize; y++ {
		row := make([]float64, imgSize)
		for x := 0; x < imgSize; x++ {
			row[x] = float64(resized[y][x])
		}
		dctRows[y] = dct.DCT1D(row)
	}

	// Take dct[:hashSize, 1:hashSize+1]: first hashSize rows, columns 1..hashSize.
	// Column 0 (DC component of each row) is skipped, matching the Python source.
	var sum float64
	block := make([][]float64, hashSize)
	for y := 0; y < hashSize; y++ {
		block[y] = make([]float64, hashSize)
		for x := 0; x < hashSize; x++ {
			v := dctRows[y][x+1]
			block[y][x] = v
			sum += v
		}
	}
	mean := sum / float64(hashSize*hashSize)

	bits := make([][]bool, hashSize)
	for y := 0; y < hashSize; y++ {
		bits[y] = make([]bool, hashSize)
		for x := 0; x < hashSize; x++ {
			bits[y][x] = block[y][x] > mean
		}
	}
	return newHashFromBits(bits), nil
}

File: hash/js/rosetta-image-hash/src/phashSimple.ts
import { Hash, ImageHashError, type RgbImage } from "./hash.js";
import { rgbToGray } from "./averageHash.js";
import { dct1d } from "./internal/dct.js";
import { toRgb } from "./internal/imgRgb.js";
import { resize } from "./internal/lanczos.js";

/**
 * phashSimple: perceptual hash using 1-D row-wise DCT, mean threshold.
 *
 * Unlike phash (which uses a 2-D DCT and median), phashSimple:
 *  1. Grayscale → Lanczos resize to (N*4, N*4).
 *  2. Apply 1-D DCT to each ROW (scipy.fftpack.dct(pixels) default axis=last=rows).
 *  3. Take top-left N rows but skip the DC column: block = dct[0:N, 1:N+1].
 *  4. avg = mean(block) as float64.
 *  5. bit = block > avg (strict >).
 *
 * This matches Python imagehash.phash_simple exactly (imagehash 4.3.2).
 */
export function phashSimple(
  img: RgbImage,
  hashSize: number,
  highfreqFactor: number = 4,
): Hash {
  if (hashSize < 2) {
    throw new ImageHashError("InvalidHashSize", `hashSize must be >= 2, got ${hashSize}`);
  }
  const imgSize = hashSize * highfreqFactor;

  const rgb = toRgb(img);
  const gray = rgbToGray(rgb.data, rgb.width, rgb.height);
  const resized = resize(gray, rgb.width, rgb.height, imgSize, imgSize);

  // Apply 1-D DCT to each row of the resized image
  // resized is row-major: row y starts at y*imgSize
  const dctRows: Float64Array[] = [];
  for (let y = 0; y < imgSize; y++) {
    const row = new Float64Array(imgSize);
    for (let x = 0; x < imgSize; x++) {
      row[x] = resized[y * imgSize + x];
    }
    dctRows.push(dct1d(row));
  }

  // Extract block: rows 0..hashSize-1, columns 1..hashSize (skip DC at col 0)
  const blockLen = hashSize * hashSize;
  const block = new Float64Array(blockLen);
  let k = 0;
  for (let y = 0; y < hashSize; y++) {
    for (let x = 1; x <= hashSize; x++) {
      block[k++] = dctRows[y][x];
    }
  }

  // Mean threshold
  let sum = 0;
  for (let i = 0; i < blockLen; i++) sum += block[i];
  const avg = sum / blockLen;

  const bits: boolean[][] = [];
  let bi = 0;
  for (let y = 0; y < hashSize; y++) {
    const row: boolean[] = new Array(hashSize);
    for (let x = 0; x < hashSize; x++) {
      row[x] = block[bi++] > avg;
    }
    bits.push(row);
  }
  return new Hash(bits);
}
